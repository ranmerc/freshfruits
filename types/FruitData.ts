export default interface FruitData {
  id: number;
  name: string;
  images: string[];
  description: string;
  shelfLife: number;
  storageInstructions: string;
  countryOfOrigin: string;
}
