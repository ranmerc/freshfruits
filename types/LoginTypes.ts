export type SuccessResponse = {
  username: string;
};

export type ErrorResponse = {
  message: string;
};

export type UserData = {
  username: string;
  email: string;
  password: string;
};

export type FormValues = {
  email: string;
  password: string;
};
