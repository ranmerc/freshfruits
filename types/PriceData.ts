export default interface PriceData {
  id: number;
  packs: Pack[];
}

export interface Pack {
  type: Type;
  quantity: number;
  price: number;
  discountPrice: number;
  inStock: boolean;
}

export enum Type {
  Count = "count",
  Weight = "weight",
}
