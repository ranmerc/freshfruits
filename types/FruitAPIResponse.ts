import FruitData from "./FruitData";

export default interface FruitAPIResponse {
  message: string;
  data: FruitData | null;
}
