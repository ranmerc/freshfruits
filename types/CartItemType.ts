import { Pack } from "./PriceData";

export default interface CartItemType {
  fruitId: number;
  selectedPackId: number;
  quantity: number;
  pack: Pack;
}
