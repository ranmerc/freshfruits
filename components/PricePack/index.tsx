import { Box, Radio, Stack, Typography } from "@mui/material";
import { Pack } from "@/types/PriceData";
import { ChangeEvent } from "react";
import calculateDiscountPercentage from "@/utils/calculateDiscountPercentage";

function PackDetails({ info, selected }: { info: Pack; selected: boolean }) {
  const discount = calculateDiscountPercentage(info.discountPrice, info.price);

  return (
    <>
      <Box>
        <Box
          sx={{
            backgroundColor: info.inStock ? "primary.main" : "gray",
            borderRadius: "8px 8px 0 0",
            textAlign: "center",
          }}
        >
          <Typography
            variant="overline"
            sx={{
              color: "primary.contrastText",
              fontWeight: "bold",
              px: 1,
            }}
          >
            {info.inStock
              ? discount
                ? `${discount}% off`
                : "Fixed Price"
              : `Out of Stock`}
          </Typography>
        </Box>

        <Stack
          sx={{
            py: 1,
            px: 2,
            border: "1px solid",
            borderColor: selected ? "secondary.main" : "inherit",
            borderRadius: "0 0 8px 8px",
            backgroundColor: selected ? "secondary.main" : "inherit",
            color: selected ? "secondary.contrastText" : "inherit",
            alignItems: "center",
          }}
        >
          <Stack direction={"row"} columnGap={0.5} alignItems={"baseline"}>
            <Typography variant="body1" fontWeight={"bold"}>
              ₹{info.discountPrice}
            </Typography>
            {!!discount && (
              <Typography
                variant="body2"
                sx={{ textDecoration: "line-through" }}
              >
                ₹{info.price}
              </Typography>
            )}
          </Stack>

          <Typography variant="caption" textAlign={"center"}>
            {info.quantity} {info.type === "count" ? "Nos" : "g"}
          </Typography>
        </Stack>
      </Box>
    </>
  );
}

export default function PricePack({
  info,
  onChange,
  selected,
  packId,
}: {
  info: Pack;
  onChange: (event: ChangeEvent<HTMLInputElement>, checked: boolean) => void;
  selected: number | null;
  packId: number;
}) {
  const currentSelected = selected === packId;

  return (
    <>
      <Radio
        checked={currentSelected}
        onChange={onChange}
        value={packId}
        name="price-packs"
        inputProps={{
          "aria-label": `${info.quantity} for ${info.discountPrice}`,
        }}
        icon={<PackDetails info={info} selected={currentSelected} />}
        checkedIcon={<PackDetails info={info} selected={currentSelected} />}
        disabled={!info.inStock}
      />
    </>
  );
}
