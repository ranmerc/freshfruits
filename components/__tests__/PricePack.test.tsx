import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import { Pack } from "@/types/PriceData";
import PricePack from "../PricePack";
import renderer from "react-test-renderer";

describe("PricePack component", () => {
  describe("Renders correct pack details", () => {
    describe("Renders correct discount", () => {
      it("Fixed Price", () => {
        const info: Pack = {
          type: "count",
          quantity: 4,
          inStock: true,
          price: 200,
          discountPrice: 200,
        };

        const component = (
          <PricePack info={info} onChange={jest.fn()} packId={1} selected={1} />
        );

        render(component);

        expect(screen.getByText(/fixed price/i)).toBeInTheDocument();

        const tree = renderer.create(component).toJSON();

        expect(tree).toMatchSnapshot();
      });

      it("Discount is 10%", () => {
        const info: Pack = {
          type: "count",
          quantity: 4,
          inStock: true,
          price: 200,
          discountPrice: 180,
        };

        const component = (
          <PricePack info={info} onChange={jest.fn()} packId={1} selected={1} />
        );

        render(component);

        expect(screen.getByText(/10% off/i)).toBeInTheDocument();

        const tree = renderer.create(component).toJSON();

        expect(tree).toMatchSnapshot();
      });
    });

    describe("Renders correct price", () => {
      it("Price on discount", () => {
        const info: Pack = {
          type: "count",
          quantity: 4,
          inStock: true,
          price: 150,
          discountPrice: 120,
        };

        const component = (
          <PricePack info={info} onChange={jest.fn()} packId={1} selected={1} />
        );

        render(component);

        expect(screen.getByText(/₹120/i)).toBeInTheDocument();
        expect(screen.getByText(/₹150/i)).toBeInTheDocument();
        expect(screen.getByText(/₹150/i)).toHaveStyle(
          "text-decoration: line-through"
        );

        const tree = renderer.create(component).toJSON();

        expect(tree).toMatchSnapshot();
      });

      it("Fixed price", () => {
        const info: Pack = {
          type: "count",
          quantity: 4,
          inStock: true,
          price: 150,
          discountPrice: 150,
        };

        const component = (
          <PricePack info={info} onChange={jest.fn()} packId={1} selected={1} />
        );

        render(component);

        expect(screen.getByText(/₹150/i)).toBeInTheDocument();
        expect(screen.getByText(/₹150/i)).not.toHaveStyle(
          "text-decoration: line-through"
        );

        const tree = renderer.create(component).toJSON();

        expect(tree).toMatchSnapshot();
      });
    });

    describe("Renders correct Quantity", () => {
      it("Renders count", () => {
        const info: Pack = {
          type: "count",
          quantity: 6,
          inStock: true,
          price: 150,
          discountPrice: 120,
        };

        const component = (
          <PricePack info={info} onChange={jest.fn()} packId={1} selected={1} />
        );

        render(component);

        expect(screen.getByText(/6 Nos/i)).toBeInTheDocument();

        const tree = renderer.create(component).toJSON();

        expect(tree).toMatchSnapshot();
      });

      it("Renders weight", () => {
        const info: Pack = {
          type: "weight",
          quantity: 300,
          inStock: true,
          price: 150,
          discountPrice: 120,
        };

        const component = (
          <PricePack info={info} onChange={jest.fn()} packId={1} selected={1} />
        );

        render(component);

        expect(screen.getByText(/300 g/i)).toBeInTheDocument();

        const tree = renderer.create(component).toJSON();

        expect(tree).toMatchSnapshot();
      });
    });
    it("Out of stock", () => {
      const info: Pack = {
        type: "count",
        quantity: 4,
        inStock: false,
        price: 200,
        discountPrice: 180,
      };

      const component = (
        <PricePack info={info} onChange={jest.fn()} packId={1} selected={1} />
      );

      render(component);

      expect(screen.getByText(/out of stock/i)).toBeInTheDocument();

      const tree = renderer.create(component).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });

  describe("Functions correctly", () => {
    it("Disabled when no stock", async () => {
      const info: Pack = {
        type: "count",
        quantity: 4,
        inStock: false,
        price: 200,
        discountPrice: 180,
      };

      render(
        <PricePack info={info} onChange={jest.fn()} packId={1} selected={2} />
      );

      expect(screen.getByRole("radio")).toBeDisabled();
    });
  });
});
