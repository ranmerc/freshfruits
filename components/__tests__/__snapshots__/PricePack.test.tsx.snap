// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PricePack component Renders correct pack details Out of stock 1`] = `
<span
  aria-disabled={true}
  className="MuiButtonBase-root MuiRadio-root MuiRadio-colorPrimary Mui-disabled PrivateSwitchBase-root MuiRadio-root MuiRadio-colorPrimary Mui-checked Mui-disabled MuiRadio-root MuiRadio-colorPrimary css-vqmohf-MuiButtonBase-root-MuiRadio-root"
  onBlur={[Function]}
  onContextMenu={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={-1}
>
  <input
    aria-label="4 for 180"
    checked={true}
    className="PrivateSwitchBase-input css-1m9pwf3"
    disabled={true}
    name="price-packs"
    onChange={[Function]}
    required={false}
    type="radio"
    value={1}
  />
  <div
    className="MuiBox-root css-0"
  >
    <div
      className="MuiBox-root css-12xkctz"
    >
      <span
        className="MuiTypography-root MuiTypography-overline css-1qfnaaq-MuiTypography-root"
      >
        Out of Stock
      </span>
    </div>
    <div
      className="MuiStack-root css-1njijdd-MuiStack-root"
    >
      <div
        className="MuiStack-root css-1qdfthb-MuiStack-root"
      >
        <p
          className="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
        >
          ₹
          180
        </p>
        <p
          className="MuiTypography-root MuiTypography-body2 css-qc7gyi-MuiTypography-root"
        >
          ₹
          200
        </p>
      </div>
      <span
        className="MuiTypography-root MuiTypography-caption css-1v2k9m3-MuiTypography-root"
      >
        4
         
        Nos
      </span>
    </div>
  </div>
</span>
`;

exports[`PricePack component Renders correct pack details Renders correct Quantity Renders count 1`] = `
<span
  className="MuiButtonBase-root MuiRadio-root MuiRadio-colorPrimary PrivateSwitchBase-root MuiRadio-root MuiRadio-colorPrimary Mui-checked MuiRadio-root MuiRadio-colorPrimary css-vqmohf-MuiButtonBase-root-MuiRadio-root"
  onBlur={[Function]}
  onContextMenu={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={null}
>
  <input
    aria-label="6 for 120"
    checked={true}
    className="PrivateSwitchBase-input css-1m9pwf3"
    disabled={false}
    name="price-packs"
    onChange={[Function]}
    required={false}
    type="radio"
    value={1}
  />
  <div
    className="MuiBox-root css-0"
  >
    <div
      className="MuiBox-root css-4f8iwy"
    >
      <span
        className="MuiTypography-root MuiTypography-overline css-1qfnaaq-MuiTypography-root"
      >
        20% off
      </span>
    </div>
    <div
      className="MuiStack-root css-1njijdd-MuiStack-root"
    >
      <div
        className="MuiStack-root css-1qdfthb-MuiStack-root"
      >
        <p
          className="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
        >
          ₹
          120
        </p>
        <p
          className="MuiTypography-root MuiTypography-body2 css-qc7gyi-MuiTypography-root"
        >
          ₹
          150
        </p>
      </div>
      <span
        className="MuiTypography-root MuiTypography-caption css-1v2k9m3-MuiTypography-root"
      >
        6
         
        Nos
      </span>
    </div>
  </div>
</span>
`;

exports[`PricePack component Renders correct pack details Renders correct Quantity Renders weight 1`] = `
<span
  className="MuiButtonBase-root MuiRadio-root MuiRadio-colorPrimary PrivateSwitchBase-root MuiRadio-root MuiRadio-colorPrimary Mui-checked MuiRadio-root MuiRadio-colorPrimary css-vqmohf-MuiButtonBase-root-MuiRadio-root"
  onBlur={[Function]}
  onContextMenu={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={null}
>
  <input
    aria-label="300 for 120"
    checked={true}
    className="PrivateSwitchBase-input css-1m9pwf3"
    disabled={false}
    name="price-packs"
    onChange={[Function]}
    required={false}
    type="radio"
    value={1}
  />
  <div
    className="MuiBox-root css-0"
  >
    <div
      className="MuiBox-root css-4f8iwy"
    >
      <span
        className="MuiTypography-root MuiTypography-overline css-1qfnaaq-MuiTypography-root"
      >
        20% off
      </span>
    </div>
    <div
      className="MuiStack-root css-1njijdd-MuiStack-root"
    >
      <div
        className="MuiStack-root css-1qdfthb-MuiStack-root"
      >
        <p
          className="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
        >
          ₹
          120
        </p>
        <p
          className="MuiTypography-root MuiTypography-body2 css-qc7gyi-MuiTypography-root"
        >
          ₹
          150
        </p>
      </div>
      <span
        className="MuiTypography-root MuiTypography-caption css-1v2k9m3-MuiTypography-root"
      >
        300
         
        g
      </span>
    </div>
  </div>
  <span
    className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
  />
</span>
`;

exports[`PricePack component Renders correct pack details Renders correct discount Discount is 10% 1`] = `
<span
  className="MuiButtonBase-root MuiRadio-root MuiRadio-colorPrimary PrivateSwitchBase-root MuiRadio-root MuiRadio-colorPrimary Mui-checked MuiRadio-root MuiRadio-colorPrimary css-vqmohf-MuiButtonBase-root-MuiRadio-root"
  onBlur={[Function]}
  onContextMenu={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={null}
>
  <input
    aria-label="4 for 180"
    checked={true}
    className="PrivateSwitchBase-input css-1m9pwf3"
    disabled={false}
    name="price-packs"
    onChange={[Function]}
    required={false}
    type="radio"
    value={1}
  />
  <div
    className="MuiBox-root css-0"
  >
    <div
      className="MuiBox-root css-4f8iwy"
    >
      <span
        className="MuiTypography-root MuiTypography-overline css-1qfnaaq-MuiTypography-root"
      >
        10% off
      </span>
    </div>
    <div
      className="MuiStack-root css-1njijdd-MuiStack-root"
    >
      <div
        className="MuiStack-root css-1qdfthb-MuiStack-root"
      >
        <p
          className="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
        >
          ₹
          180
        </p>
        <p
          className="MuiTypography-root MuiTypography-body2 css-qc7gyi-MuiTypography-root"
        >
          ₹
          200
        </p>
      </div>
      <span
        className="MuiTypography-root MuiTypography-caption css-1v2k9m3-MuiTypography-root"
      >
        4
         
        Nos
      </span>
    </div>
  </div>
  <span
    className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
  />
</span>
`;

exports[`PricePack component Renders correct pack details Renders correct discount Fixed Price 1`] = `
<span
  className="MuiButtonBase-root MuiRadio-root MuiRadio-colorPrimary PrivateSwitchBase-root MuiRadio-root MuiRadio-colorPrimary Mui-checked MuiRadio-root MuiRadio-colorPrimary css-vqmohf-MuiButtonBase-root-MuiRadio-root"
  onBlur={[Function]}
  onContextMenu={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={null}
>
  <input
    aria-label="4 for 200"
    checked={true}
    className="PrivateSwitchBase-input css-1m9pwf3"
    disabled={false}
    name="price-packs"
    onChange={[Function]}
    required={false}
    type="radio"
    value={1}
  />
  <div
    className="MuiBox-root css-0"
  >
    <div
      className="MuiBox-root css-4f8iwy"
    >
      <span
        className="MuiTypography-root MuiTypography-overline css-1qfnaaq-MuiTypography-root"
      >
        Fixed Price
      </span>
    </div>
    <div
      className="MuiStack-root css-1njijdd-MuiStack-root"
    >
      <div
        className="MuiStack-root css-1qdfthb-MuiStack-root"
      >
        <p
          className="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
        >
          ₹
          200
        </p>
      </div>
      <span
        className="MuiTypography-root MuiTypography-caption css-1v2k9m3-MuiTypography-root"
      >
        4
         
        Nos
      </span>
    </div>
  </div>
</span>
`;

exports[`PricePack component Renders correct pack details Renders correct price Fixed price 1`] = `
<span
  className="MuiButtonBase-root MuiRadio-root MuiRadio-colorPrimary PrivateSwitchBase-root MuiRadio-root MuiRadio-colorPrimary Mui-checked MuiRadio-root MuiRadio-colorPrimary css-vqmohf-MuiButtonBase-root-MuiRadio-root"
  onBlur={[Function]}
  onContextMenu={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={null}
>
  <input
    aria-label="4 for 150"
    checked={true}
    className="PrivateSwitchBase-input css-1m9pwf3"
    disabled={false}
    name="price-packs"
    onChange={[Function]}
    required={false}
    type="radio"
    value={1}
  />
  <div
    className="MuiBox-root css-0"
  >
    <div
      className="MuiBox-root css-4f8iwy"
    >
      <span
        className="MuiTypography-root MuiTypography-overline css-1qfnaaq-MuiTypography-root"
      >
        Fixed Price
      </span>
    </div>
    <div
      className="MuiStack-root css-1njijdd-MuiStack-root"
    >
      <div
        className="MuiStack-root css-1qdfthb-MuiStack-root"
      >
        <p
          className="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
        >
          ₹
          150
        </p>
      </div>
      <span
        className="MuiTypography-root MuiTypography-caption css-1v2k9m3-MuiTypography-root"
      >
        4
         
        Nos
      </span>
    </div>
  </div>
  <span
    className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
  />
</span>
`;

exports[`PricePack component Renders correct pack details Renders correct price Price on discount 1`] = `
<span
  className="MuiButtonBase-root MuiRadio-root MuiRadio-colorPrimary PrivateSwitchBase-root MuiRadio-root MuiRadio-colorPrimary Mui-checked MuiRadio-root MuiRadio-colorPrimary css-vqmohf-MuiButtonBase-root-MuiRadio-root"
  onBlur={[Function]}
  onContextMenu={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={null}
>
  <input
    aria-label="4 for 120"
    checked={true}
    className="PrivateSwitchBase-input css-1m9pwf3"
    disabled={false}
    name="price-packs"
    onChange={[Function]}
    required={false}
    type="radio"
    value={1}
  />
  <div
    className="MuiBox-root css-0"
  >
    <div
      className="MuiBox-root css-4f8iwy"
    >
      <span
        className="MuiTypography-root MuiTypography-overline css-1qfnaaq-MuiTypography-root"
      >
        20% off
      </span>
    </div>
    <div
      className="MuiStack-root css-1njijdd-MuiStack-root"
    >
      <div
        className="MuiStack-root css-1qdfthb-MuiStack-root"
      >
        <p
          className="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
        >
          ₹
          120
        </p>
        <p
          className="MuiTypography-root MuiTypography-body2 css-qc7gyi-MuiTypography-root"
        >
          ₹
          150
        </p>
      </div>
      <span
        className="MuiTypography-root MuiTypography-caption css-1v2k9m3-MuiTypography-root"
      >
        4
         
        Nos
      </span>
    </div>
  </div>
</span>
`;
